src/hotspot/share/cds/aotClassInitializer.cpp

```diff
@@ -0,0 +1,353 @@
+    //
+    // We hope we can reduce the size of this list over time, and move
+    // the responsibility for identifying such classes into the JDK
+    // code itself.  See tracking RFE JDK-8342481.
+    static AllowedSpec indy_specs[] = {
+      {"java/lang/constant/ConstantDescs"},
+      {"java/lang/constant/DynamicConstantDesc"},
+      {"java/lang/invoke/BoundMethodHandle"},
+      {"java/lang/invoke/BoundMethodHandle$Specializer"},
+      {"java/lang/invoke/BoundMethodHandle$Species_", IS_PREFIX},
+      {"java/lang/invoke/ClassSpecializer"},
+      {"java/lang/invoke/ClassSpecializer$", IS_PREFIX},
+      {"java/lang/invoke/DelegatingMethodHandle"},
+      {"java/lang/invoke/DelegatingMethodHandle$Holder"},     // UNSAFE.ensureClassInitialized()
+      {"java/lang/invoke/DirectMethodHandle"},
+      {"java/lang/invoke/DirectMethodHandle$Constructor"},
+      {"java/lang/invoke/DirectMethodHandle$Holder"},         // UNSAFE.ensureClassInitialized()
+      {"java/lang/invoke/Invokers"},
+      {"java/lang/invoke/Invokers$Holder"},                   // UNSAFE.ensureClassInitialized()
+      {"java/lang/invoke/LambdaForm"},
+      {"java/lang/invoke/LambdaForm$Holder"},                 // UNSAFE.ensureClassInitialized()
+      {"java/lang/invoke/LambdaForm$NamedFunction"},
+      {"java/lang/invoke/MethodHandle"},
+      {"java/lang/invoke/MethodHandles"},
+      {"java/lang/invoke/SimpleMethodHandle"},
+      {"java/util/Collections"},
+      {"java/util/stream/Collectors"},
+      {"jdk/internal/constant/PrimitiveClassDescImpl"},
+      {"jdk/internal/constant/ReferenceClassDescImpl"},
+
+    // Can't include this, as it will pull in MethodHandleStatics which has many environment
+    // dependencies (on system properties, etc).
+    // MethodHandleStatics is an example of a class that must NOT get the aot-init treatment,
+    // because of its strong reliance on (a) final fields which are (b) environmentally determined.
+    //{"java/lang/invoke/InvokerBytecodeGenerator"},
+
+      {nullptr}
+    };
+    if (is_allowed(indy_specs, ik)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+// TODO: currently we have a hard-coded list. We should turn this into
+// an annotation: @jdk.internal.vm.annotation.RuntimeSetupRequired
+// See JDK-8342481.
+bool AOTClassInitializer::is_runtime_setup_required(InstanceKlass* ik) {
+  return ik == vmClasses::Class_klass() ||
+         ik == vmClasses::internal_Unsafe_klass() ||
+         ik == vmClasses::ConcurrentHashMap_klass();
+}
+
+void AOTClassInitializer::call_runtime_setup(InstanceKlass* ik, TRAPS) {
+  assert(ik->has_aot_initialized_mirror(), "sanity");
+  if (ik->is_runtime_setup_required()) {
+    if (log_is_enabled(Info, cds, init)) {
+      ResourceMark rm;
+      log_info(cds, init)("Calling %s::runtimeSetup()", ik->external_name());
+    }
+    JavaValue result(T_VOID);
+    JavaCalls::call_static(&result, ik,
```

Comment:
Based on the discussions in JDK-8342283, do we need a memory fence after the call to runtimeSetup?

I think we do to be consistent with the non-AOTCache run as there is a fence after `<clinit>` (when this code would normally run) due to the initialization lock.

Reply:
I think this is already done:

`runtimeSetup()` is called inside `InstanceKlass::call_class_initializer()`. When that returns, we will proceed to `InstanceKlass::set_initialization_state_and_notify()` which will perform the memory fencing. 

https://github.com/iklam/jdk/blob/6eebd18fc2820ffb179d431f179fc6af6d1be247/src/hotspot/share/oops/instanceKlass.cpp#L1276-L1300

Reply:
I think you're right about the memory fencing being covered there but reading that code shows another potential issue - what if the `runtimeSetup()` call throws an exception?

If I read `InstanceKlass::call_class_initializer()` correctly, an exception from `runtimeSetup()` will lead to marking the class as failing initialization (`initialization_error`) which will be an unexpected outcome as there may be instances of the class in the AOTCache that aren't produced by paths through the `<clinit>` method.

Classes are able to create an instance in `<clinit>` and escape it to other code and then fail initialization.  The instances are left in an odd state where after the `<clinit>` the instance methods can be called but static fields and static methods can't be accessed.

`runtimeSetup()` expands on this corner case which I think is OK for current "crawl" stage of our "crawl-walk-run" approach as its analogous to the existing `<clinit>` behaviour.  It may need refining in the future to abort the VM or do something else (?) in the case of failed `runtimeSetup` calls

Reply:
Thanks for catching this. I added code to exit the JVM when `runtimeSetup()` fails.